bindings/pyroot/src/NumpyInterface.cxx

```diff
@@ -0,0 +1,105 @@
+// @(#)root/pyroot:$Id$
+// Author: Jim Pivarski, Jul 2017
+
+// Forwards and Python include
+#include "NumpyInterface.h"
+
+// Numpy include must be first
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#include <numpy/arrayobject.h>
+
+// ROOT
+#include <ROOT/TBulkBranchRead.hxx>
+#include <TBasket.h>
+#include <TBranch.h>
+#include <TBufferFile.h>
+#include <TBuffer.h>
+#include <TTree.h>
+
+// Bindings
+#include "PyROOT.h"
+#include "ObjectProxy.h"
+
+namespace PyROOT {
+
+  template<typename T>
+  bool getpyroot(PyObject* in, T* &out) {
+    if (!ObjectProxy_Check(in)) {
+      PyErr_Format(PyExc_TypeError, "argument must be a %s object", T::Class()->GetName());
+      return false;
+    }
+    PyROOT::ObjectProxy* objectProxy = reinterpret_cast<PyROOT::ObjectProxy*>(in);
+    
+    out = reinterpret_cast<T*>(TClass::GetClass(Cppyy::GetFinalName(objectProxy->ObjectIsA()).c_str())->DynamicCast(T::Class(), objectProxy->GetObject()));
+    
+    if (!out) {
+      PyErr_Format(PyExc_TypeError, "argument must be a %s object", T::Class()->GetName());
+      return false;
+    }
+    
+    return true;
+  }
+
+#if PY_VERSION_HEX >= 0x03000000
+  void* InitializeNumpy() {
+    import_array();
+    return 0;
+  }
+#else
+  void InitializeNumpy() {
+    import_array();
+  }
+#endif
+
+  PyObject* FillNumpyArray(PyObject* self, PyObject* args) {
```

Comment:
@bbockelm @jpivarski I wonder if we could integrate this new pythonization with the current API we have to read TTrees into numpy arrays, in particular I am talking about [AsMatrix](https://github.com/root-project/root/blob/master/bindings/pyroot/ROOT.py#L263).

`FillNumpyArray` reads a given branch with bulk I/O, while `AsMatrix` can also read a single branch of a TTree into a numpy array with regular I/O. Maybe we could think of an implementation of `AsMatrix` that supersedes `FillNumpyArray` and uses bulk I/O.